File: tests/DesktopGrass.Win2D.Tests/RainTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests;

public class RainTests
{
    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED,
                                double monitorWidth = 1920.0,
                                double density = 1.0)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, monitorWidth, density),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(seed, monitorWidth);
        sim.ResetEntities(seed);
        return sim;
    }

    private static int CountRaindrops(Sim sim) =>
        sim.Entities.Count(e => e.Kind == EntityKind.Raindrop);

    private static HashSet<uint> ObserveRaindrops(Sim sim, double seconds, double dt = 0.01)
    {
        var seen = new HashSet<uint>();
        int steps = (int)(seconds / dt);
        for (int i = 0; i < steps; i++)
        {
            sim.Tick(dt, ReadOnlySpan<InputEvent>.Empty);
            foreach (Entity e in sim.Entities)
            {
                if (e.Kind == EntityKind.Raindrop) seen.Add(e.Seed);
            }
        }
        return seen;
    }

    [Fact]
    public void EntityKindRaindropIsPinned()
    {
        Assert.Equal(0, (int)EntityKind.None);
        Assert.Equal(1, (int)EntityKind.Tumbleweed);
        Assert.Equal(2, (int)EntityKind.Snowflake);
        Assert.Equal(3, (int)EntityKind.Sheep);
        Assert.Equal(4, (int)EntityKind.Cat);
        Assert.Equal(5, (int)EntityKind.Raindrop);
        Assert.Equal(6, (int)EntityKind.Bunny);
    }

    [Fact]
    public void RaindropPrngSaltIsDistinct()
    {
        ulong[] salts =
        {
            Constants.REGROW_PRNG_SALT,
            Constants.FLOWER_PRNG_SALT,
            Constants.MUSHROOM_PRNG_SALT,
            Constants.AMBIENT_GUST_PRNG_SALT,
            Constants.CACTUS_PRNG_SALT,
            Constants.TUMBLEWEED_PRNG_SALT,
            Constants.CRITTER_PRNG_SALT,
            Constants.SNOWFLAKE_PRNG_SALT,
            Constants.PINE_PRNG_SALT,
            Constants.RAINDROP_PRNG_SALT,
        };

        Assert.Equal(salts.Length, salts.Distinct().Count());
    }

    [Fact]
    public void RainConstantsArePinned()
    {
        Assert.Equal(0xD40F0A1DD40F0A1Dul, Constants.RAINDROP_PRNG_SALT);
        Assert.Equal(6.0, Constants.RAINDROP_EMIT_RATE_PER_1920DIP);
        Assert.Equal(4.0, Constants.RAINDROP_LENGTH_MIN);
        Assert.Equal(7.0, Constants.RAINDROP_LENGTH_MAX);
        Assert.Equal(0.9, Constants.RAINDROP_THICKNESS);
        Assert.Equal(240.0, Constants.RAINDROP_FALL_SPEED_MIN);
        Assert.Equal(360.0, Constants.RAINDROP_FALL_SPEED_MAX);
        Assert.Equal(-8.0, Constants.RAINDROP_DRIFT_MIN);
        Assert.Equal(8.0, Constants.RAINDROP_DRIFT_MAX);
        Assert.Equal(0x88B0C4D0u, Constants.RAINDROP_COLOR);
        Assert.Equal(0.3, Constants.RAINDROP_LIFETIME_PADDING_SEC);
    }

    [Fact]
    public void RainDoesNotEmitInDesertOrWinter()
    {
        var sim = BuildSim();

        sim.SetScene(Scene.Desert);
        sim.Tick(5.0, ReadOnlySpan<InputEvent>.Empty);
        Assert.Equal(0, CountRaindrops(sim));

        sim.SetScene(Scene.Winter);
        sim.Tick(5.0, ReadOnlySpan<InputEvent>.Empty);
        Assert.Equal(0, CountRaindrops(sim));
    }

    [Fact]
    public void RainEmitsInGrass()
    {
        var sim = BuildSim();

        HashSet<uint> seen = ObserveRaindrops(sim, 5.0);

        Assert.True(seen.Count >= 20);
    }

    [Fact]
    public void RainEmissionRateScalesWithMonitorWidth()
    {
        var wide = BuildSim(monitorWidth: 1920.0);
        var narrow = BuildSim(monitorWidth: 960.0);

        int wideCount = ObserveRaindrops(wide, 5.0).Count;
        int narrowCount = ObserveRaindrops(narrow, 5.0).Count;

        Assert.True(wideCount > narrowCount);
        Assert.True(wideCount >= narrowCount * 1.5);
        Assert.True(wideCount <= narrowCount * 2.8);
    }

    [Fact]
    public void RaindropFieldsAreWithinExpectedRanges()
    {
        var sim = BuildSim();

        sim.TickEntities(0.0);

        Assert.True(CountRaindrops(sim) >= 1);
        foreach (Entity e in sim.Entities.Where(e => e.Kind == EntityKind.Raindrop))
        {
            Assert.InRange(e.Size, Constants.RAINDROP_LENGTH_MIN, Constants.RAINDROP_LENGTH_MAX);
            Assert.Equal(-e.Size - 2.0, e.Y, 12);
            Assert.True(e.Y < 0.0);
            Assert.InRange(e.Vy, Constants.RAINDROP_FALL_SPEED_MIN, Constants.RAINDROP_FALL_SPEED_MAX);
            Assert.InRange(e.Vx, Constants.RAINDROP_DRIFT_MIN, Constants.RAINDROP_DRIFT_MAX);
            Assert.True(e.Lifetime > 0.0);
        }
    }

    [Fact]
    public void RaindropExpiresViaLifetime()
    {
        var sim = BuildSim();
        sim.CurrentScene = Scene.Desert;
        sim.Entities.Clear();
        sim.Entities.Add(new Entity
        {
            Kind = EntityKind.Raindrop,
            Lifetime = 0.5,
            Age = 0.4,
            Vy = Constants.RAINDROP_FALL_SPEED_MIN,
        });

        sim.TickEntities(0.2);

        Assert.Equal(0, CountRaindrops(sim));
    }

    [Fact]
    public void RaindropPrngDrawOrderIsBitIdentical()
    {
        var sim = BuildSim();
        var side = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.RAINDROP_PRNG_SALT);
        double lambda = Constants.RAINDROP_EMIT_RATE_PER_1920DIP * sim.MonitorWidth / 1920.0;
        double expectedNext = 0.0;

        for (int i = 0; i < 5; i++)
        {
            sim.GlobalTime = sim.NextRaindropSpawnTime;
            sim.TickEntities(0.0);

            Assert.Equal(i + 1, CountRaindrops(sim));
            Entity e = sim.Entities[^1];
            double expectedSize = side.Uniform(Constants.RAINDROP_LENGTH_MIN, Constants.RAINDROP_LENGTH_MAX);
            double expectedX = side.Uniform(-10.0, sim.MonitorWidth + 10.0);
            double expectedFallSpeed = side.Uniform(Constants.RAINDROP_FALL_SPEED_MIN,
                                                    Constants.RAINDROP_FALL_SPEED_MAX);
            double expectedVx = side.Uniform(Constants.RAINDROP_DRIFT_MIN, Constants.RAINDROP_DRIFT_MAX);
            uint expectedSeed = side.NextU32();
            expectedNext += side.Exponential(lambda);

            Assert.Equal(expectedSize, e.Size, 12);
            Assert.Equal(expectedX, e.X, 12);
            Assert.Equal(-expectedSize - 2.0, e.Y, 12);
            Assert.Equal(expectedVx, e.Vx, 12);
            Assert.Equal(expectedFallSpeed, e.Vy, 12);
            Assert.Equal(0.0, e.Rotation, 12);
            Assert.Equal(0.0, e.RotationSpeed, 12);
            Assert.Equal(expectedSeed, e.Seed);
            Assert.Equal((sim.WindowHeight + expectedSize) / expectedFallSpeed
                         + Constants.RAINDROP_LIFETIME_PADDING_SEC, e.Lifetime, 12);
            Assert.Equal(expectedNext, sim.NextRaindropSpawnTime, 12);
        }
    }

    [Fact]
    public void SceneSwitchFromGrassToDesertSoftFadesRain()
    {
        var sim = BuildSim();
        sim.TickEntities(0.0);

        HashSet<uint> before = sim.Entities
            .Where(e => e.Kind == EntityKind.Raindrop)
            .Select(e => e.Seed)
            .ToHashSet();
        Assert.NotEmpty(before);

        sim.SetScene(Scene.Desert);
        Assert.Equal(before.Count, CountRaindrops(sim));

        sim.Tick(0.2, ReadOnlySpan<InputEvent>.Empty);
        foreach (Entity e in sim.Entities.Where(e => e.Kind == EntityKind.Raindrop))
        {
            Assert.Contains(e.Seed, before);
        }

        sim.Tick(2.0, ReadOnlySpan<InputEvent>.Empty);
        Assert.Equal(0, CountRaindrops(sim));
    }
}

File: tests/DesktopGrass.Native.Tests/src/rain_tests.cpp
#include "catch.hpp"

#include "Sim.h"
#include "Constants.h"

#include <algorithm>
#include <cstddef>
#include <iterator>
#include <set>

using namespace desktopgrass;

namespace {

Sim MakeRainTestSim(double monitorWidth = 1920.0) {
    return sim_init(CANONICAL_TEST_SEED, monitorWidth, 1.0);
}

int CountRaindrops(const Sim& sim) {
    return static_cast<int>(std::count_if(sim.entities.begin(), sim.entities.end(),
        [](const Entity& e) { return e.kind == EntityKind::Raindrop; }));
}

std::set<uint32_t> ObserveRaindrops(Sim& sim, double seconds, double dt = 0.01) {
    std::set<uint32_t> seen;
    const int steps = static_cast<int>(seconds / dt);
    for (int i = 0; i < steps; ++i) {
        sim_tick(sim, dt, nullptr, 0);
        for (const Entity& e : sim.entities) {
            if (e.kind == EntityKind::Raindrop) seen.insert(e.seed);
        }
    }
    return seen;
}

} // namespace

TEST_CASE("EntityKind::Raindrop is pinned", "[rain][enum]") {
    REQUIRE(static_cast<int>(EntityKind::None)       == 0);
    REQUIRE(static_cast<int>(EntityKind::Tumbleweed) == 1);
    REQUIRE(static_cast<int>(EntityKind::Snowflake)  == 2);
    REQUIRE(static_cast<int>(EntityKind::Sheep)      == 3);
    REQUIRE(static_cast<int>(EntityKind::Cat)        == 4);
    REQUIRE(static_cast<int>(EntityKind::Raindrop)   == 5);
}

TEST_CASE("Raindrop PRNG salt is distinct", "[rain][constants]") {
    const uint64_t salts[] = {
        REGROW_PRNG_SALT,
        FLOWER_PRNG_SALT,
        MUSHROOM_PRNG_SALT,
        AMBIENT_GUST_PRNG_SALT,
        CACTUS_PRNG_SALT,
        TUMBLEWEED_PRNG_SALT,
        CRITTER_PRNG_SALT,
        SNOWFLAKE_PRNG_SALT,
        PINE_PRNG_SALT,
        RAINDROP_PRNG_SALT,
    };
    for (std::size_t i = 0; i < std::size(salts); ++i) {
        for (std::size_t j = i + 1; j < std::size(salts); ++j) {
            REQUIRE(salts[i] != salts[j]);
        }
    }
}

TEST_CASE("Rain constants are pinned", "[rain][constants]") {
    REQUIRE(RAINDROP_PRNG_SALT == 0xD40F0A1DD40F0A1Dull);
    REQUIRE(RAINDROP_EMIT_RATE_PER_1920DIP == Approx(6.0));
    REQUIRE(RAINDROP_LENGTH_MIN == Approx(4.0));
    REQUIRE(RAINDROP_LENGTH_MAX == Approx(7.0));
    REQUIRE(RAINDROP_THICKNESS == Approx(0.9));
    REQUIRE(RAINDROP_FALL_SPEED_MIN == Approx(240.0));
    REQUIRE(RAINDROP_FALL_SPEED_MAX == Approx(360.0));
    REQUIRE(RAINDROP_DRIFT_MIN == Approx(-8.0));
    REQUIRE(RAINDROP_DRIFT_MAX == Approx(8.0));
    REQUIRE(RAINDROP_COLOR == 0x88B0C4D0u);
    REQUIRE(RAINDROP_LIFETIME_PADDING_SEC == Approx(0.3));
}

TEST_CASE("Rain does not emit in Desert or Winter", "[rain][scene]") {
    Sim sim = MakeRainTestSim();

    sim_set_scene(sim, Scene::Desert);
    sim_tick(sim, 5.0, nullptr, 0);
    REQUIRE(CountRaindrops(sim) == 0);

    sim_set_scene(sim, Scene::Winter);
    sim_tick(sim, 5.0, nullptr, 0);
    REQUIRE(CountRaindrops(sim) == 0);
}

TEST_CASE("Rain emits in Grass", "[rain][scene]") {
    Sim sim = MakeRainTestSim();

    const std::set<uint32_t> seen = ObserveRaindrops(sim, 5.0);

    REQUIRE(seen.size() >= 20);
}

TEST_CASE("Rain emission rate scales with monitor width", "[rain][rate]") {
    Sim wide = MakeRainTestSim(1920.0);
    Sim narrow = MakeRainTestSim(960.0);

    const std::size_t wideCount = ObserveRaindrops(wide, 5.0).size();
    const std::size_t narrowCount = ObserveRaindrops(narrow, 5.0).size();

    REQUIRE(wideCount > narrowCount);
    REQUIRE(static_cast<double>(wideCount) >= static_cast<double>(narrowCount) * 1.5);
    REQUIRE(static_cast<double>(wideCount) <= static_cast<double>(narrowCount) * 2.8);
}

TEST_CASE("Raindrop fields are within expected ranges", "[rain][entities]") {
    Sim sim = MakeRainTestSim();

    sim_tick_entities(sim, 0.0);

    REQUIRE(CountRaindrops(sim) >= 1);
    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Raindrop) continue;
        REQUIRE(e.size >= RAINDROP_LENGTH_MIN);
        REQUIRE(e.size <= RAINDROP_LENGTH_MAX);
        REQUIRE(e.y == Approx(-e.size - 2.0).margin(1e-12));
        REQUIRE(e.y < 0.0);
        REQUIRE(e.vy >= RAINDROP_FALL_SPEED_MIN);
        REQUIRE(e.vy <= RAINDROP_FALL_SPEED_MAX);
        REQUIRE(e.vx >= RAINDROP_DRIFT_MIN);
        REQUIRE(e.vx <= RAINDROP_DRIFT_MAX);
        REQUIRE(e.lifetime > 0.0);
    }
}

TEST_CASE("Raindrop expires via lifetime", "[rain][entities]") {
    Sim sim = MakeRainTestSim();
    sim.currentScene = Scene::Desert;
    sim.entities.clear();

    Entity e{};
    e.kind = EntityKind::Raindrop;
    e.lifetime = 0.5;
    e.age = 0.4;
    e.vy = RAINDROP_FALL_SPEED_MIN;
    sim.entities.push_back(e);

    sim_tick_entities(sim, 0.2);

    REQUIRE(CountRaindrops(sim) == 0);
}

TEST_CASE("Raindrop PRNG draw order is bit-identical", "[rain][prng]") {
    Sim sim = MakeRainTestSim();
    Prng side{};
    prng_init(side, CANONICAL_TEST_SEED ^ RAINDROP_PRNG_SALT);
    const double lambda = RAINDROP_EMIT_RATE_PER_1920DIP * sim.monitorWidth / 1920.0;
    double expectedNext = 0.0;

    for (int i = 0; i < 5; ++i) {
        sim.globalTime = sim.nextRaindropSpawnTime;
        sim_tick_entities(sim, 0.0);

        REQUIRE(CountRaindrops(sim) == i + 1);
        const Entity& e = sim.entities.back();
        const double expectedSize = prng_uniform(side, RAINDROP_LENGTH_MIN, RAINDROP_LENGTH_MAX);
        const double expectedX = prng_uniform(side, -10.0, sim.monitorWidth + 10.0);
        const double expectedFallSpeed = prng_uniform(side, RAINDROP_FALL_SPEED_MIN, RAINDROP_FALL_SPEED_MAX);
        const double expectedVx = prng_uniform(side, RAINDROP_DRIFT_MIN, RAINDROP_DRIFT_MAX);
        const uint32_t expectedSeed = prng_next_u32(side);
        expectedNext += prng_exponential(side, lambda);

        REQUIRE(e.size == Approx(expectedSize).margin(1e-12));
        REQUIRE(e.x == Approx(expectedX).margin(1e-12));
        REQUIRE(e.y == Approx(-expectedSize - 2.0).margin(1e-12));
        REQUIRE(e.vx == Approx(expectedVx).margin(1e-12));
        REQUIRE(e.vy == Approx(expectedFallSpeed).margin(1e-12));
        REQUIRE(e.rotation == Approx(0.0));
        REQUIRE(e.rotationSpeed == Approx(0.0));
        REQUIRE(e.seed == expectedSeed);
        REQUIRE(e.lifetime == Approx((sim.windowHeight + expectedSize) / expectedFallSpeed
                                      + RAINDROP_LIFETIME_PADDING_SEC).margin(1e-12));
        REQUIRE(sim.nextRaindropSpawnTime == Approx(expectedNext).margin(1e-12));
    }
}

TEST_CASE("Scene switch from Grass to Desert soft-fades rain", "[rain][scene]") {
    Sim sim = MakeRainTestSim();
    sim_tick_entities(sim, 0.0);

    std::set<uint32_t> before;
    for (const Entity& e : sim.entities) {
        if (e.kind == EntityKind::Raindrop) before.insert(e.seed);
    }
    REQUIRE_FALSE(before.empty());

    sim_set_scene(sim, Scene::Desert);
    REQUIRE(CountRaindrops(sim) == static_cast<int>(before.size()));

    sim_tick(sim, 0.2, nullptr, 0);
    for (const Entity& e : sim.entities) {
        if (e.kind == EntityKind::Raindrop) REQUIRE(before.count(e.seed) == 1);
    }

    sim_tick(sim, 2.0, nullptr, 0);
    REQUIRE(CountRaindrops(sim) == 0);
}
